tensorflow_encrypted/tensor/int32.py

```diff
@@ -135,82 +153,78 @@ def _lift(x):
-#         stacked = [j.backing[i] for j in x]
-
-#         backing.append(tf.stack(stacked, axis=axis))
-
-#     return Int100Tensor.from_decomposed(backing)
 
+class Constant(Tensor, AbstractConstant):
 
-class Constant(Tensor):
-
-    def __init__(self, value) -> None:
+    def __init__(self, value: Union[tf.Tensor, np.ndarray]) -> None:
         super(Constant, self).__init__(tf.constant(value, dtype=tf.int32))
 
-    @staticmethod
-    def from_native(value: np.ndarray) -> 'Constant':
-        assert type(value) in [np.ndarray], type(value)
-        return Constant(value)
-
-    @staticmethod
-    def from_same(value: Tensor) -> 'Constant':
-        assert type(value) in [Tensor], type(value)
-        return Constant(value.value)
-
     def __repr__(self) -> str:
         return 'int32.Constant(shape={})'.format(self.shape)
 
 
-class Placeholder(Tensor):
+class Placeholder(Tensor, AbstractPlaceholder):
 
-    def __init__(self, shape):
+    def __init__(self, shape: List[int]) -> None:
         placeholder = tf.placeholder(tf.int32, shape=shape)
         super(Placeholder, self).__init__(placeholder)
         self.placeholder = placeholder
 
-    def feed_from_native(self, value):
+    def feed_from_native(self, value: np.ndarray) -> Dict[tf.Tensor, np.ndarray]:
         assert type(value) in [np.ndarray], type(value)
         return {
             self.placeholder: value
         }
 
-    def __repr__(self):
+    def __repr__(self) -> str:
         return 'int32.Placeholder(shape={})'.format(self.shape)
 
 
-class Variable(Tensor):
+class Variable(Tensor, AbstractVariable):
 
-    def __init__(self, initial_value):
+    def __init__(self, initial_value: Union[tf.Tensor, np.ndarray]) -> None:
         variable = tf.Variable(initial_value, dtype=tf.int32, trainable=False)
         value = variable.read_value()
 
         super(Variable, self).__init__(value)
         self.variable = variable
         self.initializer = variable.initializer
 
-    @staticmethod
-    def from_native(initial_value):
-        assert type(initial_value) in [np.ndarray, tf.Tensor], type(initial_value)
-        return Variable(initial_value)
-
-    @staticmethod
-    def from_same(initial_value):
-        assert type(initial_value) in [Tensor], type(initial_value)
-        return Variable(initial_value.value)
-
-    def __repr__(self):
+    def __repr__(self) -> str:
         return 'int32.Variable(shape={})'.format(self.shape)
 
-    def assign_from_native(self, value: np.ndarray):
+    def assign_from_native(self, value: np.ndarray) -> tf.Operation:
         assert type(value) in [np.ndarray], type(value)
         return tf.assign(self.variable, value).op
 
-    def assign_from_same(self, value: Tensor):
+    def assign_from_same(self, value: Tensor) -> tf.Operation:
         assert isinstance(value, Tensor), type(value)
         return tf.assign(self.variable, value.value).op
+
+
+# type managling to make the below factory work
+Int32Tensor = Tensor
```

Comment:
maybe it's worth using the Int32 naming on all four again; I originally changed them from eg Int32Tensor to Tensor to do what you've already gone with the factory..?

Reply:
yeah sure! wasn't sure what we wanted to do/i was being lazy